src/v/iceberg/schema_protobuf.cc

```diff
@@ -0,0 +1,157 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "iceberg/schema_protobuf.h"
+
+#include "iceberg/conversion_outcome.h"
+#include "iceberg/datatypes.h"
+#include "src/v/iceberg/datatypes.h"
+
+#include <seastar/core/sstring.hh>
+
+#include <fmt/core.h>
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+#include <google/protobuf/descriptor.h>
+#include <google/protobuf/descriptor.pb.h>
+
+#include <memory>
+
+namespace iceberg {
+
+namespace {
+
+namespace pb = google::protobuf;
+
+/**
+ * Schema conversion from protocol buffers to internal Redpanda iceberg schema
+ */
+using field_outcome = conversion_outcome<nested_field_ptr>;
+using struct_outcome = conversion_outcome<struct_type>;
+
+field_outcome from_protobuf(const pb::FieldDescriptor& fd, bool is_repeated);
+
+field_outcome success(const pb::FieldDescriptor& fd, field_type ft) {
+    return nested_field::create(
+      fd.index(), fd.name(), field_required::no, std::move(ft));
+}
+
+struct_outcome struct_from_protobuf(const pb::Descriptor& msg) {
+    struct_type struct_t;
+    struct_t.fields.reserve(msg.field_count());
+    for (int i = 0; i < msg.field_count(); ++i) {
+        auto field = msg.field(i);
+        auto res = from_protobuf(*field, field->is_repeated());
+        if (res.has_error()) {
+            return res.error();
+        }
+        struct_t.fields.push_back(std::move(res).value());
+    }
+    return std::move(struct_t);
+}
+
+field_outcome from_protobuf(const pb::FieldDescriptor& fd, bool is_repeated) {
+    if (fd.is_map()) {
+        auto mt = fd.message_type();
+        auto key_field = mt->map_key();
+        auto value_field = mt->map_value();
+        auto key_nested_res = from_protobuf(
+          *key_field, key_field->is_repeated());
+        if (key_nested_res.has_error()) {
+            return key_nested_res;
+        }
+        auto value_field_nested_res = from_protobuf(
+          *value_field, value_field->is_repeated());
+        if (value_field_nested_res.has_error()) {
+            return value_field_nested_res;
+        }
+        auto field_type = map_type::create(
+          key_field->index(),
```

Comment:
Usually you never want index. That is source order used only to lookup your type in the parent.

You always want number here, which is the tag number (string name = 1;) for example